fxfs/core/api/coreapi.go

```diff
@@ -0,0 +1,204 @@
+	"context"
+	"errors"
+	"fmt"
+
+	fxfsiface "github.com/functionland/go-fula/fxfs/core/iface"
+	bserv "github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-fetcher"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	exchange "github.com/ipfs/go-ipfs-exchange-interface"
+	offlinexch "github.com/ipfs/go-ipfs-exchange-offline"
+	pin "github.com/ipfs/go-ipfs-pinner"
+	provider "github.com/ipfs/go-ipfs-provider"
+	offlineroute "github.com/ipfs/go-ipfs-routing/offline"
+	ipld "github.com/ipfs/go-ipld-format"
+	dag "github.com/ipfs/go-merkledag"
+	ipfsiface "github.com/ipfs/interface-go-ipfs-core"
+	"github.com/ipfs/interface-go-ipfs-core/options"
+	ci "github.com/libp2p/go-libp2p-core/crypto"
+	p2phost "github.com/libp2p/go-libp2p-core/host"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+	pstore "github.com/libp2p/go-libp2p-core/peerstore"
+	routing "github.com/libp2p/go-libp2p-core/routing"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	record "github.com/libp2p/go-libp2p-record"
+	madns "github.com/multiformats/go-multiaddr-dns"
+
+	"github.com/ipfs/go-namesys"
+	"github.com/ipfs/kubo/core"
+	"github.com/ipfs/kubo/core/node"
+	"github.com/ipfs/kubo/repo"
+)
+
+type CoreAPI struct {
+	nctx context.Context
+
+	identity   peer.ID
+	privateKey ci.PrivKey
+
+	repo       repo.Repo
+	blockstore blockstore.GCBlockstore
+	baseBlocks blockstore.Blockstore
+	pinning    pin.Pinner
+
+	blocks               bserv.BlockService
+	dag                  ipld.DAGService
+	ipldFetcherFactory   fetcher.Factory
+	unixFSFetcherFactory fetcher.Factory
+	peerstore            pstore.Peerstore
+	peerHost             p2phost.Host
+	recordValidator      record.Validator
+	exchange             exchange.Interface
+
+	namesys     namesys.NameSystem
+	routing     routing.Routing
+	dnsResolver *madns.Resolver
+
+	provider provider.System
+
+	pubSub *pubsub.PubSub
+
+	checkPublishAllowed func() error
+	checkOnline         func(allowOffline bool) error
+
+	// ONLY for re-applying options in WithOptions, DO NOT USE ANYWHERE ELSE
+	nd         *core.IpfsNode
+	parentOpts options.ApiSettings
+
+	// For instancing An IPFS Core API
+	nc ipfsiface.CoreAPI
+}
+
+// NewCoreAPI creates new instance of IPFS CoreAPI backed by go-ipfs Node.
+func NewCoreAPI(n *core.IpfsNode, c ipfsiface.CoreAPI, opts ...options.ApiOption) (fxfsiface.CoreAPI, error) {
+	parentOpts, err := options.ApiOptions()
+	if err != nil {
+		return nil, err
+	}
+	return (&CoreAPI{nd: n, parentOpts: *parentOpts, nc: c}).WithOptions(opts...)
+}
+
+// Unixfs returns the UnixfsAPI interface implementation backed by the go-ipfs node
+func (api *CoreAPI) PrivateFS() fxfsiface.PrivateFS {
+	return (*PrivateAPI)(api)
+}
+
+// Block returns the BlockAPI interface implementation backed by the go-ipfs node
+func (api *CoreAPI) PublicFS() fxfsiface.PublicFS {
+	return (*PublicAPI)(api)
+}
+
+// WithOptions returns api with global options applied
+func (api *CoreAPI) WithOptions(opts ...options.ApiOption) (fxfsiface.CoreAPI, error) {
```

Comment:
We might want to take these in the constructor. Also change the function name to make it clear these are IPFS options.

Reply:
done